text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1796 @@
+
+This RFC proposes mechanisms that programmers will use to confidently answer such questions—by checking whether the `TransmuteFrom` and `TransmuteInto` traits are implemented, or (equivalently) by checking whether the `can_transmute` predicate (a `const fn`) is satisfied.
+
+## Expressing Library Layout Guarantees
+There is no canonical way for crate authors to declare the SemVer layout guarantees of their types. Crate authors currently must state their layout guarantees using prose in their documentation. In contrast to structural stability (e.g., the declared visibility of fields), layout stability is expressed extra-linguistically.
+
+This isn't satisfactory: guarantees expressed in prose outside of the Rust programming language are guarantees that cannot be reasoned about *inside* the language. Whereas `rustc` can dutifully deny programmers access to private fields, it is unable to prevent programmers from making unfounded expectations of types' in-memory layouts.
+
+This RFC proposes simple-but-powerful [mechanisms][stability] for declaring layout stability guarantees.
+
+## Expressing Layout Requirements
+Similarly, there is no canonical way for crate authors to declare the layout requirements of generic abstractions over types that have certain layout properties. 
+
+For instance, a common bit-packing technique involves using the relationship between allocations and alignment. If a type is aligned to 2<sup>n</sup>, then the *n* least significant bits of pointers to that type will equal `0`. These known-zero bits can be packed with data. Since alignment cannot be currently reasoned about at the type-level, it's currently impossible to bound instantiations of a generic parameter based on minimum alignment.
+
+The mechanisms proposed by the RFC enable this, see [here][0000-ext-layout-traits.md].
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Terminology & Concepts
+
+### 📖 Transmutation
+**Transmutation** is the act of reinterpreting the bytes corresponding to a value of one type as if they corresponded to a different type. Concretely, we mean the behavior of this function:
+```rust
+#[inline(always)]
+unsafe fn transmute<Src, Dst>(src: Src) -> Dst
+{
+    #[repr(C)]
+    union Transmute<Src, Dst> {
+        src: ManuallyDrop<Src>,
+        dst: ManuallyDrop<Dst>,
+    }
+
+    ManuallyDrop::into_inner(Transmute { src: ManuallyDrop::new(src) }.dst)
+}
+```
+
+### 📖 Safer Transmutation
+By **safer transmutation** we mean: *what `where` bound could we add to `transmute` restricts its type parameters `Src` and `Dst` in ways that statically limit the function's misuse?* Our answer to this question will ensure that transmutations are, by default, *sound*, *safe* and *stable*.
+
+### 📖 Soundness
+A transmutation is ***sound*** if the mere act of transmuting a value from one type to another is not unspecified or undefined behavior.
```

Comment:
Ah, I must have overlooked this... while the terminology here is self-consistent within the file, it is not consistent with the wider ecosystem. 
Specifically, see [the UCG definition of soundness](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#soundness-of-code--of-a-library), which corresponds to what you call "safety".

If you'd ask me to define safety I'd ask "safety of what". "safety of code" IMO means "no `unsafe {}`" (Rustaceans say things like "you can do this in safe code"); "safety of a program execution" IMO means that there was no UB.

I thought the UCG's use of "soundness" corresponds to the usual consensus, but maybe not -- I'd be interested in feedback. For this RFC, the odd one is the notion you call "soundness"; I don't think I have seen a good name for it before -- I would have called it "not UB"/"UB-free". Maybe "well-defined" would work -- the transmute itself has well-defined semantics, but unless it is sound (in the UCG sense) it can still cause UB later.

Reply:
> I would have called it "not UB"/"UB-free".

That isn't quite right though. Both "soundness" and "safety" as described in this RFC are about avoiding undefined behavior. The distinction is where that happens. "soundness" is about whether undefined behavior happens at the transmutation itself, whereas "safety" is about whether undefined behavior could occur at some later point after the transmutation (for example attempting to dereference an invalid pointer).

Reply:
> That isn't quite right though. Both "soundness" and "safety" as described in this RFC are about avoiding undefined behavior. The distinction is where that happens. 

Yes, I am aware. A (in your terminology) sound-but-not-safe transmute *is not UB* itself. That's why I think it makes sense to call it "well-defined" or "UB-free".

Usually, "sound" means "can be called from safe code in arbitrary ways without causing UB", so my proposal is certainly not more wrong than what the RFC says. ;)

> "soundness" is about whether undefined behavior happens at the transmutation itself, whereas "safety" is about whether undefined behavior could occur at some later point after the transmutation (for example attempting to dereference an invalid pointer).

I understand that that's how the RFC uses those words. However, I do not think that's how anyone else uses those words -- I certainly have not seen this sue of "sound"/"safe" before (even though I am involve in many discussions around this topic), and the UCG glossary above represents consensus of the UCG WG, so quite a few people doing unsafe-related work agree.

Reply:
In fact it's not just the UCG, even [the Rust reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) agrees with me (well, I write that text, but it passed review):

> It is the programmer's responsibility when writing unsafe code to ensure that any safe code interacting with the unsafe code cannot trigger these behaviors. unsafe code that satisfies this property for any safe client is called sound; if unsafe code can be misused by safe code to exhibit undefined behavior, it is unsound.

Reply:
I'm not saying that this usage of the RFC's usage of soundness/safeness is right, just that I'm not sure "UB-free" is any better or less confusing. TBH, I'm not entirely sure what the value of distinguishing between these two kinds of safety is.

Reply:
> I'm not saying that this usage of the RFC's usage of soundness/safeness is right, just that I'm not sure "UB-free" is any better or less confusing. 

Fair. I think I personally prefer "well-defined" -- in the sense that a UB transmute literally has no meaning. A transmute that is not itself UB but might cause UB later *has* meaning (the Rust Abstract Machine says what happens when you run it); it is only later that the program could become meaningless.

Unsafe code frequently works with things that are not UB but would be UB if arbitrary safe code could no run. In [this old blogpost](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html) I call such data "valid but not safe". I think that is basically the same distinction that the RFC is making. So I think the distinction is fundamental, but [there is no consensus on how exactly to call it](https://github.com/rust-lang/unsafe-code-guidelines/issues/95).

Reply:
I like the term "well defined". I'll switch to it in my next round of edits.